src/librados/librados_asio.h

```diff
@@ -0,0 +1,294 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2017 Red Hat, Inc.
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ */
+
+#ifndef LIBRADOS_ASIO_H
+#define LIBRADOS_ASIO_H
+
+#include <memory>
+#include <boost/asio.hpp>
+#include "include/rados/librados.hpp"
+#include "common/backport_std.h"
+
+/// Defines asynchronous librados operations that satisfy all of the
+/// "Requirements on asynchronous operations" imposed by the C++ Networking TS
+/// in section 13.2.7. Many of the type and variable names below are taken
+/// directly from those requirements.
+///
+/// The current draft of the Networking TS (as of 2017-11-27) is available here:
+/// http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/n4711.pdf
+///
+/// The boost::asio documentation duplicates these requirements here:
+/// http://www.boost.org/doc/libs/1_66_0/doc/html/boost_asio/reference/asynchronous_operations.html
+
+namespace librados {
+
+namespace detail {
+
+/// unique_ptr with custom deleter for AioCompletion
+struct release_completion {
+  void operator()(AioCompletion *c) { c->release(); }
+};
+using unique_completion_ptr =
+    std::unique_ptr<AioCompletion, release_completion>;
+
+/// Invokes the given completion handler. When the type of Result is not void,
+/// storage is provided for it and that result is passed as an additional
+/// argument to the handler.
+template <typename Result>
+struct invoker {
+  Result result;
+  template <typename CompletionHandler>
+  void invoke(CompletionHandler& completion_handler,
+              boost::system::error_code ec) {
```

Comment:
Do we still need to use boost::system::error_code with the upgraded boost? Or is there a way to have it use the C++ <system_error> facility instead?

Reply:
good question. as far as i know, we're stuck with `boost::system::error_code` for now. a grep for `std::error_code` in the asio tree came up empty. the boost docs have a [Networking TS compatibility](http://www.boost.org/doc/libs/1_66_0/doc/html/boost_asio/net_ts.html) section, but there's no mention of `std::error_code` there either
> the interfaces themselves are free of boost types

so it looks like i lied here :(